### DISCUSSION TITLE
List of all possible "tags"?

### DISCUSSION THREAD

**Discussion by Linardos** [FETS-AI/Challenge#72] [Q&A]
In the notebook it is specified that:
![image](https://user-images.githubusercontent.com/26694607/125072894-dc0cc100-e0c3-11eb-9b5f-aafd0a5c1400.png)

There are examples of what a tag might be here but I couldn't find a comprehensive list of all the possible tags. You mention that the same 'tensor_name' and 'round'  are used in such a case with the tag 'trained' to specify the weights after training. What would I use instead of "trained" to obtain the initial weights for example? Is there a list detailing all possible tags that I missed somewhere?

**Comment by sarthakpati** [comment]
Pinging @msheller @alexey-gruzdev @psfoley

**Reply by alexey-gruzdev** [reply]
@Linardos hi! Thanks for the interest! Let me provide an example of possible state of db for [PyTorch MNIST example](https://github.com/intel/openfl/blob/develop/openfl-tutorials/Federated_Pytorch_MNIST_Tutorial.ipynb). I believe based on this info, it would be helpful to understand the mechanics better. The attached file is a dataframe after 5 rounds of PyTorch MNIST example with db_store_rounds=5. You would have an access to all of this except of tags with starts with ('metric', _, _). They are not allowed for custom aggregation. For them we use simple averaging, since there is no much sense to aggregate them in a different manner.

**Reply by Linardos** [reply]
Thanks for the response. Unfortunately I'm a bit confused on how this relates to my question. I can't see the db_iterator being defined anywhere in the PyTorch MNIST example so I find it hard to correlate it to the Challenge notebook we are provided with. Maybe I'm missing something? I basically want to understand what are all the possible values of the "tags" part for the db_iterator. In one of the provided examples I can see "aggregated" and "delta" 
![image](https://user-images.githubusercontent.com/26694607/125174364-abb14980-e1cd-11eb-84be-0e02331b2e6c.png)
What else is there? Is it possible to obtain an intermediate model value within the same round using this parameter? (For example, I set models to train for 5 epochs per round, can I obtain the weights for epochs 3 and 4 somehow?)

**Reply by alexey-gruzdev** [reply]
@Linardos  I meant that network was from PyTorch MNIST example. Maybe I put a wrong link here and that's what confuse you.  So, [here](https://github.com/intel/openfl/blob/develop/openfl-tutorials/Federated_Pytorch_MNIST_custom_aggregation_Tutorial.ipynb)  you could see how we work with db_iterator to implement different schemes.

**Reply by alexey-gruzdev** [reply]
@Linardos I am sorry again, looks like in [original answer](https://github.com/FETS-AI/Challenge/discussions/72#discussioncomment-985420) for some reason attachment didn't work. I would put here a screen of dataframe that I got from MNIST. Hope it helps!

![image](https://user-images.githubusercontent.com/77294631/125472998-c150756d-a79a-4a77-a962-a5244a3e2a5f.png)

**Reply by psfoley** [reply]
@Linardos to follow onto what Alexey provided, there is a flow that the tags follow in the normal training sequence. At the highest level, the

`('model',)` -> These are the tags given to the aggregated tensor that will be downloaded from the aggregator to the collaborator for the purpose of training or aggregated_model_validation. 

 `('trained',)` -> As you correctly pointed out, the 'trained' tag is used to denote the trained model weights coming from each collaborator. There is a transformation that happens to the tensorkey as it is sent from the collaborator to the aggregator, namely that the 'origin' of the tensorkey changes from the common name of the collaborator (let's use as example collaborator name `one`). The reason for this is that the tensor is decompressed once it moves to the aggregator, and then because that tensor can only be found on the aggregator, it's origin becomes `aggregator`. The collaborator name is then moved into the tag, which is where you see `('trained','one',)` in Alexey's image above.
 
 `('aggregated',)` -> This tag is the result of applying a built-in or custom aggregation scheme to the collaborator model weights.  It should be noted that this is **not** the final version that will be downloaded by the collaborator at the start of the next round. This `('aggregated',)` tensor then undergoes compression and decompression, at which point the tensor is then relabeled `('model,')`. The reason for this is to guarantee that the version of the model downloaded by the collaborator is the same as what is stored on the aggregator (i.e. for lossy compression of model weights, which is not used in this competition)
 
 `(*,'delta',) -> The `delta` tag is an option provided within OpenFL to only send the difference between model weights between the aggregator and collaborator. This can be very useful for lossy compression. This flag is [not enabled](https://github.com/FETS-AI/Challenge/blob/main/Task_1/openfl-workspace/fets_challenge_workspace/plan/plan.yaml#L22) for the competition, so you don't need to worry about this tag.

**Reply by GorkemP** [reply]
Dear organizers,

I am still confused about tags and it prevents us implementing our method. When I inspect the details of db_iterator, I see that for any layer, value of  `(trained,)` at round x-1 equals to value of `(aggregated,)` at round x. So, it seems to me that after aggragating the models coming from collaborators, resulting aggregated model layer/weights are tagged as `(trained,)` (for that round x-1) and `(aggregated,)` is the model distributed to collaborators at the start of the next round. For example, at round 3, if I want to get model weights that are distributed to collaborators, I think I should look for `(aggregated,)` at round 3 not round 2. But, example codes contradict with this, as @Linardos shared above (as screen-shot), previous round is fl_round-1 not fl_round. What am I missing?

**Reply by psfoley** [reply]
@GorkemP, I'm sorry for the confusion. Your understanding is correct. The example @Linardos posted above is the way that you could resolve the **prior round's** aggregated tensor.  

![image](https://user-images.githubusercontent.com/3803396/126014668-72970fbd-b37a-400a-a20e-0d0ec2186c46.png)

At the point when you have access to the db_iterator, the aggregated tensors have not been generated for the current round. If it is round 1 for example, you would be able to see each of the `('trained',*)` tensors produced by each of the collaborators performing the training task for round1 , but the output of your custom aggregation function will produce the `('aggregated',)` tensor for round 2. Using the `('aggregated',)` tensor for round 1 would be the layer value the collaborator would have started with for the current round prior to training. Does this make sense?

**Reply by GorkemP** [reply]
@psfoley Thank you for the answer. It totally makes sense, we can only get aggregated result in the next round (with `(aggregated,)` tag for round x or `(trained,)` tag for round x-1 ). I still think (or not understand) that this example is confusing:

![image](https://user-images.githubusercontent.com/7123842/126057592-fec74f1c-c85c-4d26-bd82-30877feafa5b.png)

Since it uses `(aggregated,)` tag for  `(fl_round-1)`, `previos_tensor_value` refers to **two** step before. Yet, we normally refer gradients as single steps, this is the confusing part for me. In our aggregation method, we should calculate gradients for each collaborator with respect to the last aggregated model (we should measure how much each collaborator training changed the weights), in our case I believe that I should just change the code from `fl_round-1` to `fl_round` in the above example.
